cloud/log_syncer.go

```diff
@@ -0,0 +1,203 @@
+// Copyright 2023 Terramate GmbH
+// SPDX-License-Identifier: MPL-2.0
+
+package cloud
+
+import (
+	"bufio"
+	"bytes"
+	"io"
+	"sync"
+	"time"
+
+	"github.com/rs/zerolog/log"
+	"github.com/terramate-io/terramate/errors"
+)
+
+type (
+	// LogSyncer is the log syncer controller type.
+	LogSyncer struct {
+		pending      DeploymentLogs
+		fds          []io.Closer
+		in           chan *DeploymentLog
+		lastEnqueued time.Time
+		syncfn       Syncer
+		wg           sync.WaitGroup
+		shutdown     chan struct{}
+
+		maxLineSize  int
+		batchSize    int
+		idleDuration time.Duration
+	}
+
+	// Syncer is the actual synchronizer callback.
+	Syncer func(l DeploymentLogs)
+)
+
+// DefaultLogMaxLineSize is the default maximum line.
+// TODO(i4k): to be removed.
+const DefaultLogMaxLineSize = 4096
+
+// DefaultLogBatchSize is the default batch size.
+const DefaultLogBatchSize = 256
+
+// DefaultLogIdleDuration is the maximum idle duration before a sync could happen.
+const DefaultLogIdleDuration = 1 * time.Second
+
+// NewLogSyncer creates a new log syncer.
+func NewLogSyncer(syncfn Syncer) *LogSyncer {
+	return NewLogSyncerWith(syncfn, DefaultLogMaxLineSize, DefaultLogBatchSize, DefaultLogIdleDuration)
+}
+
+// NewLogSyncerWith creates a new customizable syncer.
+func NewLogSyncerWith(
+	syncfn Syncer,
+	maxLineSize int,
+	batchSize int,
+	idleDuration time.Duration,
+) *LogSyncer {
+	if maxLineSize == 0 {
+		panic("max line size must be set")
+	}
+	l := &LogSyncer{
+		in:       make(chan *DeploymentLog, batchSize),
```

Comment:
I think the buffer size of this channel should be independent of the batch size? The consumer takes values out as soon as they are available and adds them to pending, so there isn't really any relation to the batch size.

Reply:
yes but I feel that there's a good relationship between the batch size and the channel throughput, but just subjective at this point. As discussed before, this value only affects performance, as it could be even zero (unbuffered) without any issues other than performance.
I will come back to it later, in upcoming PRs.

Thanks @snakster